packages/hdwallet-metamask-multichain/src/adapter.ts

```diff
@@ -0,0 +1,68 @@
+import MetaMaskOnboarding from "@metamask/onboarding";
+import * as core from "@shapeshiftoss/hdwallet-core";
+import { enableShapeShiftSnap, shapeShiftSnapInstalled } from "@shapeshiftoss/metamask-snaps-adapter";
+import { createStore } from "mipd";
+
+import { SNAP_ID } from "./common";
+import { MetaMaskMultiChainHDWallet } from "./shapeshift-multichain";
+
+export const mipdstore = createStore();
+
+export class MetaMaskAdapter {
+  keyring: core.Keyring;
+  providerRdns: string;
+
+  private constructor(keyring: core.Keyring, providerRdns: string) {
+    this.keyring = keyring;
+    this.providerRdns = providerRdns;
+  }
+
+  public static useKeyring(keyring: core.Keyring, providerRdns: string) {
+    return new MetaMaskAdapter(keyring, providerRdns);
+  }
+
+  public async initialize(): Promise<number> {
+    return Object.keys(this.keyring.wallets).length;
+  }
+
+  public async pairDevice(): Promise<MetaMaskMultiChainHDWallet | undefined> {
+    const maybeEip6963Provider = mipdstore.findProvider({ rdns: this.providerRdns });
+    if (!maybeEip6963Provider) {
```

Comment:
we don't want to trigger the metamask onboarding if trying to pair a non MM wallet (xdefi/coinbase/tally/etc). otherwise just generic throw.
`if (!maybeEip6963Provider && isMetaMask) {`

Reply:
e0e8f4e5

Reply:
We want to still throw a generic error on `!maybeEip6963Provider`?

Reply:
Absolutely - as discussed this wasn't really a concern previously (if we ever hit this, we could safely assume there *is* a provider since its rdns was originally detected in web) but since we're now adding a static list of mipd providers (regardless of installation), then we absolutely want to throw here in case we hit a static non-MetaMask rdns (Coinbase, XDEFI, Rabby) and it isn't installed: 1b4ad64b